ENH Check for nonfinite in mlp models

chritter opened the pull request:
#### Reference Issues/PRs

Addresses #17925.

#### What does this implement/fix? Explain your changes.

This implements check for non-finite values of estimator parameters: as produced from optimizers in MLP models.

#### Any other comments?

Work done with @norbusan.

Originally part of larger PR #22092

chritter committed:
non-finite check in mlp params.

Co-authored-by: Norbert Preining <norbert@preining.info>

chritter committed:
added changelog entry

chritter committed:
updated changelog

thomasjpfan commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -440,6 +440,20 @@ def _fit(self, X, y, incremental=False):
+            lambda coefs, intercept: np.isfinite(coefs).all()
+            and np.isfinite(intercept).all()
+        )
+        if not np.all(
+            [
+                check_weights(coef, inter)
+                for coef, inter in zip(self.coefs_, self.intercepts_)
+            ]
+        ):
```
We can simplify the approach here:

```python
from itertools import chain

weights = chain(self.coefs_, self.intercepts_)
if not all(np.isfinite(w).all() for w in weights):
    ...
```

thomasjpfan commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -440,6 +440,20 @@ def _fit(self, X, y, incremental=False):
+                for coef, inter in zip(self.coefs_, self.intercepts_)
+            ]
+        ):
+            raise ValueError("Solver produced non-finite parameter weights.")
```
Similar comment from https://github.com/scikit-learn/scikit-learn/pull/22149/files#r780722680 We can adjust the message so users know what they can do about it.

```suggestion
            raise ValueError("coefs or intercepts have non-finite parameter weights. "
                             "The input data may contain large values and need to be "
                             "preprocessed.")
```

thomasjpfan replied:
May you include the second part of the message?

thomasjpfan replied:
May you include the second part of the error message?

thomasjpfan reviewed (commented):
Thank you for the PR @chritter !

chritter committed:
simplied nonfinite param check and test;

test_nonfinite_params test data now generated; test parameter in mlp with itertools

Co-authored-by: Norbert Preining <norbert@preining.info>

chritter commented:
Great suggestions @thomasjpfan which I adopted. Thank you!

chritter committed:
reformatting test file

chritter committed:
update changelog formatting to pass circleci check

thomasjpfan commented on doc/whats_new/v1.1.rst:
```diff
@@ -462,6 +462,15 @@ Changelog
   left corner of the HTML representation to show how the elements are
   clickable. :pr:`21298` by `Thomas Fan`_.
 
+:mod:`sklearn.neural_network`
+.............................
```
Nit: May we move this header up such that the modules are in alphabetical order?

thomasjpfan reviewed (commented):
Thanks for the update!

thomasjpfan reviewed (commented):
Thanks for the update!

chritter committed:
update changelog location alphabetic, extended error message and test

thomasjpfan commented on doc/whats_new/v1.1.rst:
```diff
@@ -462,6 +470,7 @@ Changelog
   left corner of the HTML representation to show how the elements are
   clickable. :pr:`21298` by `Thomas Fan`_.
 
+
```
This change is unrelated. Can you revert?

chritter replied:
@thomasjpfan This line belongs to the enhancement changelog entry which was not part of this PR changes:
```
- |Enhancement| :func:`utils.estimator_html_repr` displays an arrow on the top
  left corner of the HTML representation to show how the elements are
  clickable. :pr:`21298` by `Thomas Fan`_.
```

thomasjpfan replied:
There is a blank line here

thomasjpfan replied:
Maybe it's the GitHub interface. I am referring to this newline added by this PR:

![Screen Shot 2022-01-17 at 9 02 06 AM](https://user-images.githubusercontent.com/5402633/149782516-cf6835f7-b48f-43e2-b9e5-caaa1f7f3958.jpg)

chritter replied:
Thanks for the clarification.

thomasjpfan reviewed (approved):
I left a minor comment, otherwise LGTM!

jjerphan commented on doc/whats_new/v1.1.rst:
```diff
@@ -393,6 +393,14 @@ Changelog
+:mod:`sklearn.neural_network`
+.............................
+
+- |Enhancement| :func:`neural_network.MLPClassifier` and 
+  :func:`neural_network.MLPRegressor` shows error
+  messages when optimizers produce non-finite parameter weights. :pr:`22150`
+  by :user:`Christian Ritter <chritter>` and :user:`Norbert Preining <norbusan>`.
```
```suggestion
- |Enhancement| :func:`neural_network.MLPClassifier` and 
  :func:`neural_network.MLPRegressor` show error
  messages when optimizers produce non-finite parameter weights. :pr:`22150`
  by :user:`Christian Ritter <chritter>` and :user:`Norbert Preining <norbusan>`.
```

jjerphan commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -506,6 +506,24 @@ def test_partial_fit_errors():
 
 
+def test_nonfinite_params():
+    # Check MLPRegressor throws ValueError when dealing with non-finite
+    # parameter values
```
```suggestion
    # Check that MLPRegressor throws ValueError when dealing with non-finite
    # parameter values
```

jjerphan reviewed (approved):
LGTM after addressing @thomasjpfan's comments.

chritter committed:
removed empty line in whatsnew

chritter committed:
Apply suggestions from code review

improved changelog description

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
